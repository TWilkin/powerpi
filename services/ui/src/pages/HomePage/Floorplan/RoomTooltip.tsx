import { Sensor } from "@powerpi/common-api";
import classNames from "classnames";
import { useMemo } from "react";
import { Fragment } from "react/jsx-runtime";
import BatteryIcon from "../../../components/BatteryIcon";
import SensorIcon from "../../../components/SensorIcon";
import Tooltip from "../../../components/Tooltip";
import Value from "../../../components/Value";

type RoomTooltipProps = {
    name: string;

    floor: string;

    room: string;

    sensors: Sensor[];
};

const RoomTooltip = ({ name, floor, room, sensors }: RoomTooltipProps) => {
    const showingBattery = useMemo(
        () => sensors.findIndex((sensor) => (sensor.batterySince ?? 0) > 0) !== -1,
        [sensors],
    );

    return (
        <Tooltip id={generateRoomTooltipId(floor, room)} place="top">
            <h3>{name}</h3>

            <div
                className={classNames("grid auto-rows-auto", {
                    "grid-cols-[20px_20px_auto_1fr]": showingBattery,
                    "grid-cols-[20px_auto_1fr]": !showingBattery,
                })}
            >
                {sensors.map((sensor) => (
                    <Fragment key={sensor.name}>
                        <SensorIcon type={sensor.type} state={sensor.state} />

                        {showingBattery &&
                            (sensor.battery != null ? <BatteryIcon device={sensor} /> : <span />)}

                        <p className="px-2">{sensor.display_name}</p>

                        <p>
                            <Value type={sensor.type} value={sensor.value} unit={sensor.unit} />
                        </p>
                    </Fragment>
                ))}
            </div>
        </Tooltip>
    );
};
export default RoomTooltip;

export function generateRoomTooltipId(floor: string, room: string) {
    return `room-tooltip-${floor}-${room}`;
}
