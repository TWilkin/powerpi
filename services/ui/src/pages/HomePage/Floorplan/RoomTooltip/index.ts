export { default as generateRoomTooltipId } from "./generateRoomTooltipId";
export { default } from "./RoomTooltip";
