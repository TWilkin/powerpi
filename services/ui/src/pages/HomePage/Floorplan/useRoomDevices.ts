import { BaseDevice, DeviceState } from "@powerpi/common-api";
import { useMemo } from "react";
import { chain as _ } from "underscore";
import useQueryDevices from "../../../queries/useQueryDevices";
import useQuerySensors from "../../../queries/useQuerySensors";

type DeviceList =
    | {
          deviceType: "device" | "sensor";
          device: BaseDevice;
          state: DeviceState | string | undefined;
      }[]
    | undefined;

/** Hook to get, and organise by count and type all the devices/sensors in a room on the floorplan. */
export default function useRoomDevices(room: string) {
    const { data: devices } = useQueryDevices();
    const { data: sensors } = useQuerySensors();

    return useMemo(() => {
        const deviceList: DeviceList = devices.map((device) => ({
            deviceType: "device",
            device,
            state: device.state,
        }));

        const sensorList: DeviceList = sensors.map((sensor) => ({
            deviceType: "sensor",
            device: sensor,
            state: sensor.state,
        }));

        // get all the devices and sensors and group by type, then get the count
        const all = _(deviceList)
            .concat(sensorList)
            .filter(({ device }) => device.location === room && device.visible)
            .groupBy(({ deviceType, device }) => `${deviceType}_${device.type}`)
            .map((group) => {
                const states = _(group)
                    .map((device) => device.state)
                    .unique()
                    .value();

                return {
                    deviceType: group[0].deviceType,
                    type: group[0].device.type,
                    count: group.length,
                    state: states.length === 1 ? states[0] : undefined,
                };
            });

        // sort by count descending, then type alphabetically
        return all
            .sortBy((device) => device.type)
            .reverse()
            .sortBy((device) => device.count)
            .reverse()
            .value();
    }, [devices, room, sensors]);
}
