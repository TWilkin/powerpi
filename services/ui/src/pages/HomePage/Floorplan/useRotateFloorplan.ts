import { Floor, Floorplan, Point, Room } from "@powerpi/common-api";
import { useMemo } from "react";
import { isPolygonRoom, isRectangleRoom, pointsFromRect, viewBoxByFloorplan } from "./ViewBox";

export default function useRotateFloorplan(floorplan: Floorplan, rotate: boolean) {
    return useMemo(() => {
        const viewBox = viewBoxByFloorplan(floorplan);

        if (!rotate) {
            return {
                floorplan,
                viewBox,
            };
        }

        const radians = degreesToRadians(90);

        const centre = {
            x: (viewBox.maxX - viewBox.minX) / 2,
            y: (viewBox.maxY - viewBox.minY) / 2,
        };

        const rotatedFloorplan = rotateFloorplan(floorplan, radians, centre);

        const rotatedViewBox = viewBoxByFloorplan(rotatedFloorplan);

        if (rotatedViewBox.minX !== 0 || rotatedViewBox.minY !== 0) {
            const offset = {
                x: -rotatedViewBox.minX,
                y: -rotatedViewBox.minY,
            };

            const translatedFloorplan = translateFloorplan(rotatedFloorplan, offset);

            const translatedViewBox = viewBoxByFloorplan(translatedFloorplan);

            return {
                floorplan: translatedFloorplan,
                viewBox: translatedViewBox,
            };
        }

        return {
            floorplan: rotatedFloorplan,
            viewBox: rotatedViewBox,
        };
    }, [floorplan, rotate]);
}

function degreesToRadians(degrees: number) {
    return (degrees * Math.PI) / 180;
}

function rotateFloorplan(floorplan: Floorplan, radians: number, centre: Point): Floorplan {
    const floors = floorplan.floors.map((floor) => rotateFloor(floor, radians, centre));

    return {
        ...floorplan,
        floors,
    };
}

function rotateFloor(floor: Floor, radians: number, centre: Point): Floor {
    const rooms = floor.rooms.map((room) => rotateRoom(room, radians, centre));

    return {
        ...floor,
        rooms,
    };
}

function rotateRoom(room: Room, radians: number, centre: Point): Room {
    // if it's a rectangle convert to points for convenience
    let points = isPolygonRoom(room)
        ? room.points
        : pointsFromRect(room.x, room.y, room.width, room.height);

    // now everything is a polygon so just rotate it
    points = points.map((point) => rotate(point, radians, centre));

    // if it was a rectangle we want to convert it back
    if (points.length === 4) {
        const x = Math.min(...points.map((point) => point.x));
        const y = Math.min(...points.map((point) => point.y));
        const width = Math.max(...points.map((point) => point.x)) - x;
        const height = Math.max(...points.map((point) => point.y)) - y;

        return {
            ...room,
            points: undefined,
            x,
            y,
            width,
            height,
        };
    }

    return {
        ...room,
        points,
    };
}

function rotate(point: Point, radians: number, centre: Point): Point {
    const sin = Math.sin(radians);
    const cos = Math.cos(radians);

    return {
        x: cos * (point.x - centre.x) - sin * (point.y - centre.y) + centre.x,
        y: sin * (point.x - centre.x) - cos * (point.y - centre.y) + centre.y,
    };
}

function translateFloorplan(floorplan: Floorplan, offset: Point): Floorplan {
    const floors = floorplan.floors.map((floor) => translateFloor(floor, offset));

    return {
        ...floorplan,
        floors,
    };
}

function translateFloor(floor: Floor, offset: Point): Floor {
    const rooms = floor.rooms.map((room) => translateRoom(room, offset));

    return {
        ...floor,
        rooms,
    };
}

function translateRoom(room: Room, offset: Point): Room {
    if (isRectangleRoom(room)) {
        const point = { x: room.x ?? 0, y: room.y ?? 0 };
        const translated = translate(point, offset);

        return {
            ...room,
            x: translated.x,
            y: translated.y,
        };
    }

    return {
        ...room,
        points: room.points.map((point) => translate(point, offset)),
    };
}

function translate(point: Point, offset: Point): Point {
    return {
        x: point.x + offset.x,
        y: point.y + offset.y,
    };
}
