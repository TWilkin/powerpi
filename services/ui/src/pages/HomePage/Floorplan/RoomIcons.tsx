import { Room } from "@powerpi/common-api";
import { useCallback, useMemo } from "react";
import { useTranslation } from "react-i18next";
import { chain as _ } from "underscore";
import DeviceIcon from "../../../components/DeviceIcon";
import Icon from "../../../components/Icon";
import SensorIcon from "../../../components/SensorIcon";
import { isPolygonRoom } from "./ViewBox";
import useRoomDevices from "./useRoomDevices";

type RoomIconsProps = {
    room: Room;
};

/** Component to display device/sensor icons in a room on the floorplan. */
const RoomIcons = ({ room }: RoomIconsProps) => {
    const { t } = useTranslation();

    const devices = useRoomDevices(room.name);

    const { scale, iconPaddedSize, deviceCount, iconsWide, offsetX, offsetY } = useMemo(() => {
        const iconSize = 40;
        const iconPadding = 4;
        const iconPaddedSize = iconSize + 2 * iconPadding;

        const scale = 1 / (1024 / iconSize);

        // calculate the room size
        let x = room.x ?? 0;
        let y = room.y ?? 0;
        let width = 0;
        let height = 0;

        // this could put the icons outside the polygon so needs to be improved
        if (isPolygonRoom(room)) {
            x = Math.min(...room.points.map((point) => point.x));
            y = Math.min(...room.points.map((point) => point.y));

            width = Math.max(...room.points.map((point) => point.x)) - x;
            height = Math.max(...room.points.map((point) => point.y)) - y;
        } else {
            width = room.width ?? 0;
            height = room.height ?? 0;
        }

        // next we need to find the centre of the room
        const centreX = width / 2 + x;
        const centreY = height / 2 + y;

        // how many icons can we fit?
        let iconsWide = Math.floor(width / iconPaddedSize);
        let iconsTall = Math.floor(height / iconPaddedSize);

        // we want to leave some space either side
        iconsWide = Math.max(iconsWide - 2, 0);
        iconsTall = Math.max(iconsTall - 2, 0);

        // limit them to what is necessary to centre everything
        const best = bestValue(devices.length, iconsWide, iconsTall);
        if (best) {
            iconsWide = best[0];
            iconsTall = best[1];
        }

        // how many devices can we actually show
        let deviceCount = Math.min(devices.length, iconsWide * iconsTall);
        if (deviceCount < devices.length) {
            // if we have too many devices we'll show an ellipsis instead of the last one
            deviceCount--;
        }

        // now we can work out the offsets
        const offsetX = centreX - (iconsWide * iconPaddedSize) / 2;
        const offsetY = centreY - (iconsTall * iconPaddedSize) / 2;

        return {
            scale,
            iconPaddedSize,
            deviceCount,
            iconsWide,
            offsetX,
            offsetY,
        };
    }, [devices.length, room]);

    const getTransform = useCallback(
        (index: number) => {
            // find out which row and column it's in
            let row = 0;
            while (index >= (row + 1) * iconsWide) {
                row++;
            }
            const column = index - row * iconsWide;

            // now we can work out the position
            const x = offsetX + column * iconPaddedSize;
            const y = offsetY + row * iconPaddedSize;

            return `translate(${x} ${y}) scale(${scale})`;
        },
        [iconPaddedSize, iconsWide, offsetX, offsetY, scale],
    );

    return (
        <>
            {_(devices)
                .first(deviceCount)
                .map((device, i) => (
                    <g key={`${device.deviceType}_${device.type}`} transform={getTransform(i)}>
                        <title>
                            {device.type} ({device.count})
                        </title>

                        {device.deviceType === "device" && <DeviceIcon type={device.type} />}

                        {device.deviceType === "sensor" && (
                            <SensorIcon type={device.type} state={device.state} />
                        )}
                    </g>
                ))
                .value()}

            {deviceCount < devices.length && (
                <g transform={getTransform(deviceCount)}>
                    <title>
                        {t("pages.home.more devices", { count: devices.length - deviceCount })}
                    </title>

                    <Icon icon="more" />
                </g>
            )}
        </>
    );
};
export default RoomIcons;

/** Generate factors for value starting with the values closes to the square root. */
function* factors(value: number) {
    const range = Math.floor(Math.sqrt(value));

    for (let i = range; i >= 1; i--) {
        if (value % i === 0) {
            yield [i, value / i];
        }
    }
}

/** Generate factors between count and rows * columns, in case the factors of count won't fit. */
function* increasingFactors(count: number, rows: number, columns: number) {
    const maxCount = Math.min(count, rows * columns);

    for (let i = maxCount; i <= rows * columns; i++) {
        for (const f of factors(i)) {
            yield f;
        }
    }
}

/** Generate factors of count that will work for the rows and columns. */
function* feasibleFactors(count: number, rows: number, columns: number) {
    for (const [f1, f2] of increasingFactors(count, rows, columns)) {
        if (f1 <= columns && f2 <= rows) {
            yield [f2, f1];
        } else if (f2 <= columns && f1 <= rows) {
            yield [f1, f2];
        }
    }
}

/** Find the "best" factors of count in a grid of no more than rows x columns. */
function bestValue(count: number, rows: number, columns: number) {
    for (const factors of feasibleFactors(count, rows, columns)) {
        return factors;
    }

    return undefined;
}
