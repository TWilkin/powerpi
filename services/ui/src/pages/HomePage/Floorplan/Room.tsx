import { Room as IRoom } from "@powerpi/common-api";
import classNames from "classnames";
import { useMemo } from "react";
import RoomIcons from "./RoomIcons";
import { generateRoomTooltipId } from "./RoomTooltip";
import { isPolygonRoom, isRectangleRoom } from "./ViewBox";

type RoomProps = {
    floor: string;

    room: IRoom;
};

const shapeClasses = classNames(
    "pointer-events-all",
    "fill-none outline-none",
    "stroke-text",
    "hover:fill-bg-hover",
);

/** Component representing a room in the home floorplan. */
const Room = ({ floor, room }: RoomProps) => {
    const points = useMemo(() => {
        if (isPolygonRoom(room)) {
            return room.points.map((point) => `${point.x},${point.y}`).join(" ");
        }

        return undefined;
    }, [room]);

    return (
        <g className="focus:outline-none" data-tooltip-id={generateRoomTooltipId(floor, room.name)}>
            {isRectangleRoom(room) && (
                <rect
                    x={room.x}
                    y={room.y}
                    width={room.width}
                    height={room.height}
                    className={shapeClasses}
                />
            )}

            {points && <polygon points={points} className={shapeClasses} />}

            <RoomIcons room={room} />
        </g>
    );
};
export default Room;
