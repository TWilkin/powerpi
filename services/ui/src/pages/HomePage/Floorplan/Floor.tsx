import { Floor as IFloor } from "@powerpi/common-api";
import Room from "./Room";

type FloorProps = {
    floor: IFloor;
};

/** Component representing a floor in the home floorplan. */
const Floor = ({ floor }: FloorProps) => (
    <g>
        {floor.rooms.map((room) => (
            <Room key={room.name} room={room} />
        ))}
    </g>
);
export default Floor;
