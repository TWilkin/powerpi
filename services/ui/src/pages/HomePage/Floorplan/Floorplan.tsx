import { Floorplan as IFloorplan } from "@powerpi/common-api";
import { useMemo } from "react";
import useFloor from "../useFloor";
import Floor from "./Floor";

type FloorplanProps = {
    floorplan: IFloorplan;
};

/** Component representing the SVG of the home floorplan. */
const Floorplan = ({ floorplan }: FloorplanProps) => {
    const currentFloorName = useFloor();

    const currentFloor = useMemo(
        () => floorplan.floors.find((floor) => floor.name === currentFloorName),
        [currentFloorName, floorplan.floors],
    );

    return (
        <>
            {currentFloor && (
                <svg preserveAspectRatio="xMidYMid" className="flex-1">
                    <Floor floor={currentFloor} />
                </svg>
            )}
        </>
    );
};
export default Floorplan;
