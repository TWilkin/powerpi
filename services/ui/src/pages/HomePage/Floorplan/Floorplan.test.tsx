import { DeviceState, Floorplan as IFloorplan } from "@powerpi/common-api";
import { render } from "@testing-library/react";
import { vi } from "vitest";
import Floorplan from "./Floorplan";

const mocks = vi.hoisted(() => ({
    useOrientation: vi.fn(),
    useQueryDevices: vi.fn(),
    useQuerySensors: vi.fn(),
    useFloor: vi.fn(),
}));

vi.mock("../../../hooks/useOrientation", () => ({ default: mocks.useOrientation }));
vi.mock("../../../queries/useQueryDevices", () => ({ default: mocks.useQueryDevices }));
vi.mock("../../../queries/useQuerySensors", () => ({ default: mocks.useQuerySensors }));
vi.mock("../useFloor", () => ({ default: mocks.useFloor }));

const devices = [
    createDevice("Socket1", "socket", "LivingRoom"),
    createDevice("Socket2", "socket", "LivingRoom"),
    createDevice("Light1", "light", "LivingRoom"),
    createDevice("Light2", "light", "MasterBedroom"),
    createDevice("Light3", "light"),
];

const sensors = [
    createDevice("Temperature1", "temperature", "LivingRoom"),
    createDevice("Temperature2", "temperature", "MasterBedroom"),
];

describe("Floorplan", () => {
    const floorplan: IFloorplan = {
        floors: [
            {
                name: "Ground",
                rooms: [
                    {
                        name: "LivingRoom",
                        display_name: "Living Room",
                        width: 100,
                        height: 100,
                    },
                    { name: "Kitchen", width: 100, height: 100, x: 100 },
                ],
            },
            {
                name: "First",
                rooms: [
                    {
                        name: "MasterBedroom",
                        display_name: "Master Bedroom",
                        width: 200,
                        height: 100,
                    },
                ],
            },
        ],
    };

    beforeEach(() => {
        mocks.useOrientation.mockReturnValue({ isLandscape: true, isPortrait: false });
        mocks.useQueryDevices.mockReturnValue({
            data: devices,
        });
        mocks.useQuerySensors.mockReturnValue({ data: sensors });

        mocks.useFloor.mockReturnValue("Ground");
    });

    test("renders ground floor", () => {
        mocks.useFloor.mockReturnValue("Ground");

        const { container } = render(<Floorplan floorplan={floorplan} />);

        expect(container).toMatchSnapshot();
    });

    test("renders first floor", () => {
        mocks.useFloor.mockReturnValue("First");

        const { container } = render(<Floorplan floorplan={floorplan} />);

        expect(container).toMatchSnapshot();
    });

    test("renders portrait", () => {
        mocks.useOrientation.mockReturnValue({ isLandscape: false, isPortrait: true });

        const { container } = render(<Floorplan floorplan={floorplan} />);

        expect(container).toMatchSnapshot();
    });
});

function createDevice(name: string, type: string, location: string = "Some Room", visible = true) {
    return {
        name,
        display_name: name,
        type,
        location,
        state: DeviceState.Unknown,
        since: 1,
        visible,
    };
}
