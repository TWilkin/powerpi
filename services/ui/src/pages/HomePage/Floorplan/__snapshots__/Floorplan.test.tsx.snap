// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Floorplan > renders first floor 1`] = `
<div>
  <div
    class="relative flex-1"
  >
    <svg
      class="absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 w-[95%] h-[95%]"
      preserveAspectRatio="xMidYMid"
      viewBox="0 0 200 100"
    >
      <g>
        <g
          class="focus:outline-none"
          data-tooltip-id="room-tooltip-First-MasterBedroom"
        >
          <rect
            class="pointer-events-all fill-none outline-none stroke-text hover:fill-bg-hover"
            height="100"
            width="200"
          />
          <g
            transform="translate(4 50) scale(0.0390625)"
          >
            <title>
              There are 3 more hidden devices/sensors.
            </title>
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-ellipsis "
              data-icon="ellipsis"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 256a56 56 0 1 1 112 0A56 56 0 1 1 8 256zm160 0a56 56 0 1 1 112 0 56 56 0 1 1 -112 0zm216-56a56 56 0 1 1 0 112 56 56 0 1 1 0-112z"
                fill="currentColor"
              />
            </svg>
          </g>
        </g>
      </g>
    </svg>
  </div>
</div>
`;

exports[`Floorplan > renders ground floor 1`] = `
<div>
  <div
    class="relative flex-1"
  >
    <svg
      class="absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 w-[95%] h-[95%]"
      preserveAspectRatio="xMidYMid"
      viewBox="0 0 200 100"
    >
      <g>
        <g
          class="focus:outline-none"
          data-tooltip-id="room-tooltip-Ground-LivingRoom"
        >
          <rect
            class="pointer-events-all fill-none outline-none stroke-text hover:fill-bg-hover"
            height="100"
            width="100"
          />
          <g
            transform="translate(2 50) scale(0.0390625)"
          >
            <title>
              There are 4 more hidden devices/sensors.
            </title>
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-ellipsis "
              data-icon="ellipsis"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 256a56 56 0 1 1 112 0A56 56 0 1 1 8 256zm160 0a56 56 0 1 1 112 0 56 56 0 1 1 -112 0zm216-56a56 56 0 1 1 0 112 56 56 0 1 1 0-112z"
                fill="currentColor"
              />
            </svg>
          </g>
        </g>
        <g
          class="focus:outline-none"
          data-tooltip-id="room-tooltip-Ground-Kitchen"
        >
          <rect
            class="pointer-events-all fill-none outline-none stroke-text hover:fill-bg-hover"
            height="100"
            width="100"
            x="100"
          />
        </g>
      </g>
    </svg>
  </div>
</div>
`;

exports[`Floorplan > renders portrait 1`] = `
<div>
  <div
    class="relative flex-1"
  >
    <svg
      class="absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 w-[95%] h-[95%]"
      preserveAspectRatio="xMidYMid"
      viewBox="0 0 100 200"
    >
      <g>
        <g
          class="focus:outline-none"
          data-tooltip-id="room-tooltip-Ground-LivingRoom"
        >
          <rect
            class="pointer-events-all fill-none outline-none stroke-text hover:fill-bg-hover"
            height="100"
            width="100"
            x="0"
            y="0"
          />
          <g
            transform="translate(2 50) scale(0.0390625)"
          >
            <title>
              There are 4 more hidden devices/sensors.
            </title>
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-ellipsis "
              data-icon="ellipsis"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 256a56 56 0 1 1 112 0A56 56 0 1 1 8 256zm160 0a56 56 0 1 1 112 0 56 56 0 1 1 -112 0zm216-56a56 56 0 1 1 0 112 56 56 0 1 1 0-112z"
                fill="currentColor"
              />
            </svg>
          </g>
        </g>
        <g
          class="focus:outline-none"
          data-tooltip-id="room-tooltip-Ground-Kitchen"
        >
          <rect
            class="pointer-events-all fill-none outline-none stroke-text hover:fill-bg-hover"
            height="100"
            width="100"
            x="7.105427357601002e-15"
            y="100"
          />
        </g>
      </g>
    </svg>
  </div>
</div>
`;
